neqo-crypto/src/constants.rs

```diff
@@ -6,21 +6,60 @@
 
 #![allow(dead_code)]
 
+use enum_map::Enum;
+
 use crate::ssl;
 
 // Ideally all of these would be enums, but size matters and we need to allow
 // for values outside of those that are defined here.
 
 pub type Alert = u8;
 
-pub type Epoch = u16;
-// TLS doesn't really have an "initial" concept that maps to QUIC so directly,
-// but this should be clear enough.
-pub const TLS_EPOCH_INITIAL: Epoch = 0_u16;
-pub const TLS_EPOCH_ZERO_RTT: Epoch = 1_u16;
-pub const TLS_EPOCH_HANDSHAKE: Epoch = 2_u16;
-// Also, we don't use TLS epochs > 3.
-pub const TLS_EPOCH_APPLICATION_DATA: Epoch = 3_u16;
+#[derive(Default, Debug, Enum)]
+pub enum Epoch {
+    // TLS doesn't really have an "initial" concept that maps to QUIC so directly,
+    // but this should be clear enough.
+    #[default]
+    Initial = 0,
+    ZeroRtt,
+    Handshake,
+    ApplicationData,
+    // Also, we don't use TLS epochs > 3.
+}
+
+impl From<u16> for Epoch {
+    fn from(e: u16) -> Self {
+        match e {
+            0 => Self::Initial,
+            1 => Self::ZeroRtt,
+            2 => Self::Handshake,
+            3 => Self::ApplicationData,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl From<Epoch> for usize {
+    fn from(e: Epoch) -> Self {
+        match e {
+            Epoch::Initial => 0,
+            Epoch::ZeroRtt => 1,
+            Epoch::Handshake => 2,
+            Epoch::ApplicationData => 3,
+        }
+    }
+}
+
+impl std::fmt::Display for Epoch {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Initial => write!(f, "Initial"),
+            Self::ZeroRtt => write!(f, "ZeroRtt"),
+            Self::Handshake => write!(f, "Handshake"),
+            Self::ApplicationData => write!(f, "ApplicationData"),
+        }
+    }
+}
```

Comment:
Why do we need a `Display` implementation?

Reply:
Good catch! We used `{}` in a `qdebug!`. Changed that to `{:?}`.